cmd/ttn-lw-cli/commands/applications_pubsub.go

```diff
@@ -40,16 +42,44 @@ func applicationPubSubIDFlags() *pflag.FlagSet {
 	return flagSet
 }
 
-func natsFlags() *pflag.FlagSet {
+func applicationPubSubProviderFlags() *pflag.FlagSet {
 	flagSet := &pflag.FlagSet{}
 	flagSet.Bool("nats", false, "use the NATS provider")
-	flagSet.String("nats-server-url", "", "")
+	flagSet.AddFlagSet(natsProviderApplicationPubSubFlags)
+	flagSet.Bool("mqtt", false, "use the MQTT provider")
+	flagSet.AddFlagSet(mqttProviderApplicationPubSubFlags)
+	flagSet.AddFlagSet(dataFlags("mqtt.tls-ca", ""))
+	flagSet.AddFlagSet(dataFlags("mqtt.tls-client-cert", ""))
+	flagSet.AddFlagSet(dataFlags("mqtt.tls-client-key", ""))
+	addDeprecatedProviderFlags(flagSet)
 	return flagSet
 }
 
+func addDeprecatedProviderFlags(flagSet *pflag.FlagSet) {
+	util.DeprecateFlag(flagSet, "nats_server_url", "nats.server_url")
+}
+
+func forwardDeprecatedProviderFlags(flagSet *pflag.FlagSet) {
+	util.ForwardFlag(flagSet, "nats_server_url", "nats.server_url")
+}
+
+func forwardProviderDataFlags(flagSet *pflag.FlagSet) error {
```

Comment:
This forwards the contents of the certificates / private keys `local-file` flags to the flags that would expect the input file as `hex`. 

Example: `--mqtt.tls-ca` is the flag that expects the CA PEM-encoded certificate to be provided as a hex string, while `--mqtt.tls-ca-local-file` is a location of the CA certificate. The `forward*` functions move the data from the file into the actual flags that expect the contents. Why do it like this instead of simply having only the data-flags ? I considered this would be a bit more sane since the string-expecting flags are already introduced through reflection, and it keeps the code generic for any future providers.

Reply:
OK I understand. The idea of the data flags, however, is that we can add sources later, without changing the commands or doing any forwarding magic. Now, the data flags only support local files, but we could add a flag for hex or remote files by setting them in `dataFlags()`. In any case, `getDataBytes()` and `getDataReader()` simply look at which flags are available and load the data accordingly.

So: if you want to add support for hex, do it in `dataFlags()` by introducing a `[-]hex` counterpart of `[-]local-file`. But, I don't think we need this at all, people are not going to pass hex through flags and presumably neither through env

Reply:
I've fully dropped the forwarding magic and changed it to only using `getDataBytes`.